packages/react-components/react-avatar/src/components/AvatarGroup/useAvatarGroup.tsx

```diff
@@ -0,0 +1,117 @@
+import { MoreHorizontalRegular } from '@fluentui/react-icons';
+import { PopoverSurface } from '@fluentui/react-popover';
+import { useId } from '@fluentui/react-utilities';
+import type { AvatarGroupProps, AvatarGroupState } from './AvatarGroup.types';
+
+/**
+ * Create the state required to render AvatarGroup.
+ *
+ * The returned state can be modified with hooks such as useAvatarGroupStyles_unstable,
+ * before being passed to renderAvatarGroup_unstable.
+ *
+ * @param props - props from this instance of AvatarGroup
+ * @param ref - reference to root HTMLElement of AvatarGroup
+ */
+export const useAvatarGroup_unstable = (props: AvatarGroupProps, ref: React.Ref<HTMLElement>): AvatarGroupState => {
+  const {
+    children,
+    layout = 'spread',
+    maxAvatars = 5,
+    overflowIndicator = 'number-overflowed',
+    size = 32,
+    ...rest
+  } = props;
+  const id = useId('avatarGroup-', props.id);
+  const childrenCount = React.Children.count(children);
+  const numOfAvatarsToShow = layout === 'pie' ? 3 : maxAvatars;
+  const hasOverflow = childrenCount > numOfAvatarsToShow;
+  const childrenArray = React.Children.toArray(children);
+
+  const rootChildren = childrenArray.slice(0, numOfAvatarsToShow).map(child => {
+    if (React.isValidElement(child)) {
+      // If the layout is pie and the size is less than 40, the Avatar should only display the first initial
+      let initials = child.props.initials;
+
+      if (layout === 'pie' && child.props.name && size < 40) {
+        // No need to check if it's rtl since we will use only the first letter.
+        initials = getInitials(child.props.name, false)[0];
+      }
+
+      // Overwritting size to the one given in the props.
+      return React.cloneElement(child, { size: size, initials: initials, color: child.props.color ?? 'colorful' });
+    }
+    return child;
+  });
+
+  const root = getNativeElementProps('div', {
+    ref,
+    id: id,
+    role: 'group',
+    children: rootChildren,
+    ...rest,
+  });
+
+  const popoverTriggerChildren =
+    layout === 'pie' || overflowIndicator === 'icon' ? null : `+${childrenCount - numOfAvatarsToShow}`;
+  const popoverTriggerIcon = layout !== 'pie' && overflowIndicator === 'icon' ? <MoreHorizontalRegular /> : undefined;
+
+  const popoverTrigger = resolveShorthand(props.popoverTrigger, {
+    required: true,
+    defaultProps: {
+      children: popoverTriggerChildren,
+      shape: 'circular',
+      icon: popoverTriggerIcon,
+      appearance: layout === 'pie' ? 'transparent' : 'outline',
+    },
+  });
+
+  const popoverChildren =
+    hasOverflow &&
+    childrenArray.slice(numOfAvatarsToShow).map((child, k) => {
+      if (React.isValidElement(child)) {
+        // Avatars inside PopoverSurface must be size 24
+        return (
+          <li className={extraAvatarGroupClassNames.popoverSurfaceItem} key={k} tabIndex={0}>
+            {React.cloneElement(child, { size: 24, color: child.props.color ?? 'colorful' })}
```

Comment:
Ditto above about messing with Avatar props here. Could this be done via a context instead?